File: 백준/Silver/11726. 2×n 타일링/2×n 타일링.cc
#include <bits/stdc++.h>
using namespace std;

int n;
int cache[1005];


int tiling(int cur){
    if(cur <= 1) return 1;

    int &ret = cache[cur];
    if(ret != -1 ) return ret;

    return ret = (tiling(cur-1) + tiling(cur-2)) %10007;
}

int main(){
    cin >> n;
    memset(cache, -1, sizeof(cache));
    cout << tiling(n);
    return 0;
}

File: 백준/Silver/11726. 2×n 타일링/2×n 타일링.py
import sys
sys.setrecursionlimit(10000000)



def tiling(n):
    if n <= 1 : return 1

    if cache[n] != -1 : return cache[n]

    cache[n] = (tiling(n-1) + tiling(n-2)) %10007
    return cache[n]

n = int(input())
cache = [-1 for i in range(n+1)]



print(tiling(n))